### PR TITLE
Notifier efficiency

### LINKED ISSUE #1198: Notifier panics at start
### Description of Problem / Feature Request
When deploying notifier to OCP, it panics at the start. Full log is attached. I'm not gonna go into great detail as this problem is already addressed in #1193. I just created this issue to point out that this can be observed also on the `main` branch and is not unique for Hank's `notifier-efficiency` branch.
[clair-notifier-1-59g25.log](https://github.com/quay/clair/files/6100519/clair-notifier-1-59g25.log)

### Environment
Encountered when running clair from `main` branch, commit 102ae88dd84c1f769b8c037226d92b301d887aab.

### PR DESCRIPTION
**Pull request by hdonnay**
This changeset attempts to improve memory usage across clair, and especially around the `affected_manifests` call.

* It attempts to use `io.Pipe` to cross-wire JSON encoding and API requests, to avoid buffering the entire body request in memory.
* It replaces usage of `encoding/json` with `github.com/ugorji/go/codec` configured for JSON in order to allow streaming the JSON encoding.
* It chunks `affected_manifests` calls in the notifier. This should prevent large vulnerability turnovers from causing extremely large API calls.

cc @jan-zmeskal

### CODE REVIEW

**Comment by hdonnay** [comment]
@ldelossa I'm unsure if the composed `AffectedManifest` objects need a pass to deduplicate and sort the `VulnerableManifests` member.

**Comment by hdonnay** [comment]
Undrafting because I added a `Sort` call.

File: notifier/processor.go [hunk 1]
```diff
@@ -126,24 +127,35 @@ func (p *Processor) create(ctx context.Context, e Event, prev uuid.UUID) error {
...
-		Vulnerabilities:     make(map[string]*claircore.Vulnerability),
-		VulnerableManifests: make(map[string][]string),
+	tab := notifTab{
+		N:      make([]Notification, 0),
+		lookup: make(map[string]int),
 	}
 	eg, wctx := errgroup.WithContext(ctx)
-	eg.Go(getAffected(wctx, p.indexer, diff.Added, added))
-	eg.Go(getAffected(wctx, p.indexer, diff.Removed, removed))
+	eg.Go(getAffected(wctx, p.indexer, !p.NoSummary, diff.Added, Added, &tab))
```
Comment (ldelossa) [inline notifier/processor.go]: Could you keep the same meaning of "NoSummary" all the way down, even into the getAffected method, instead of negating the bool here?

File: notifier/processor.go [hunk 2]
```diff
@@ -213,10 +191,16 @@ func min(a, b int) int {
 	return b
 }
 
+type notifTab struct {
```
Comment (ldelossa) [inline notifier/processor.go:197]: comment the purpose of this.

File: httptransport/client/indexer.go [hunk 3]
```diff
@@ -13,43 +12,50 @@ import (
...
 		return nil, err
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode != http.StatusOK {
-		return nil, &clairerror.ErrRequestFail{Code: resp.StatusCode, Status: resp.Status}
-	}
-	err = json.NewDecoder(resp.Body).Decode(&affected)
-	if err != nil {
-		return nil, &clairerror.ErrBadAffectedManifests{err}
+		return nil, &clairerror.ErrRequestFail{
```
Comment (ldelossa) [inline httptransport/client/indexer.go:41]: Why the change in error type?

File: httptransport/client/indexer.go [hunk 4]
```diff
@@ -59,36 +65,40 @@ func (s *HTTP) AffectedManifests(ctx context.Context, v []claircore.Vulnerabilit
...
+	var ir claircore.IndexReport
+	switch ct := resp.Header.Get("content-type"); ct {
+	case "", `application/json`:
+		dec := codec.GetDecoder(resp.Body)
+		defer codec.PutDecoder(dec)
+		if err := dec.Decode(&ir); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unrecognized content-type %q", ct)
```
Comment (ldelossa) [inline httptransport/client/indexer.go:99]: I don't know if forcing the client to provide the json content type is that useful. 

We export a JSON api, we should try to parse json, regardless of whether a client is providing a content type or not. I think we ran into this already and it was an annoyance more then a feature.

File: httptransport/client/matcher.go [hunk 5]
```diff
@@ -25,31 +25,33 @@ func (c *HTTP) Scan(ctx context.Context, ir *claircore.IndexReport) (*claircore.
...
-		return nil, err
+	switch ct := req.Header.Get("content-type"); ct {
+	case "", `application/json`:
+		dec := codec.GetDecoder(resp.Body)
+		defer codec.PutDecoder(dec)
+		if err := dec.Decode(&vr); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unrecognized content-type %q", ct)
```
Comment (ldelossa) [inline httptransport/client/matcher.go:54]: same, don't think this should be a requirement.

File: httptransport/notificationshandler.go [hunk 6]
```diff
@@ -149,9 +146,8 @@ func (h *NotifHandler) Get(w http.ResponseWriter, r *http.Request) {
 		Notifications: notifications,
 	}
 
-	err = json.NewEncoder(w).Encode(&response)
-	if err != nil {
-		log.Error().Err(err).Msg("failed to serialize response")
-	}
-	return
+	defer writerError(w, &err)()
```
Comment (ldelossa) [inline httptransport/notificationshandler.go:149]: Still confused about whether this is too useful, how is anything going to know to check a trailer?

File: httptransport/updateoperationhandler.go [hunk 7]
```diff
@@ -114,7 +115,5 @@ func (h *UOHandler) Delete(w http.ResponseWriter, r *http.Request) {
 			Message: fmt.Sprintf("could not get update operations: %v", err),
 		}
 		je.Error(w, resp, http.StatusInternalServerError)
-		return
```
Comment (ldelossa) [inline httptransport/updateoperationhandler.go:117]: is this a linter thing?

File: internal/codec/codec_test.go [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+func Example() {
+	enc := GetEncoder(os.Stdout)
+	defer PutEncoder(enc)
+	enc.MustEncode([]string{"a", "slice", "of", "strings"})
+	fmt.Fprintln(os.Stdout)
+	enc.MustEncode(nil)
+	fmt.Fprintln(os.Stdout)
+	enc.MustEncode(map[string]string{})
+	fmt.Fprintln(os.Stdout)
+	// Output: ["a","slice","of","strings"]
```
Comment (ldelossa) [inline internal/codec/codec_test.go:22]: probably want to test with json large then your 4096 buffer size to make sure things are OK.

File: notifier/callback.go [hunk 9]
```diff
@@ -12,40 +13,39 @@ import (
...
+// MarshalJSON implements json.Marshaler.
 func (cb Callback) MarshalJSON() ([]byte, error) {
-	var m = map[string]string{
-		"notification_id": cb.NotificationID.String(),
-		"callback":        cb.Callback.String(),
-	}
-	return json.Marshal(m)
+	// This writes out a block of bytes directly, instead of allocating a
+	// cbMessage structure, serializing the URL, then copying it into a []byte.
+	b := bytes.NewBufferString(`{"callback":"`)
```
Comment (ldelossa) [inline notifier/callback.go]: this looks like a rough thing to maintain if the schema changes. 
is this strictly necessary?

File: notifier/processor.go [hunk 10]
```diff
@@ -124,17 +126,36 @@ func (p *Processor) create(ctx context.Context, e Event, prev uuid.UUID) error {
...
+				removed++
+			}
+		}
+		ev.
+			Int("added", added).
+			Int("removed", removed).
+			Msg("affected manifest counts")
+	}
+
+	if len(tab.N) == 0 {
```
Comment (ldelossa) [inline notifier/processor.go:158]: why is this condition added? You're still processing all potential notifications in this method call right?

**Review by ldelossa — CHANGES_REQUESTED** [review]
I'm not 100% sure the pipe driven JSONReader should be used across the board. There are resource considerations to initializing the channels it uses internally. It may not be necessary to use them where memory constraints are not a concern.

File: notifier/processor.go (hunk 10 above)
Comment (hdonnay) [reply notifier/processor.go:158]: It's the equivalent to `if len(added.VulnerableManifests) == 0 && len(removed.VulnerableManifests) == 0 {`

File: httptransport/client/indexer.go (hunk 3 above)
Comment (hdonnay) [reply httptransport/client/indexer.go:41]: I don't know what you mean, it's still `ErrRequestFail`.

File: httptransport/client/indexer.go (hunk 4 above)
Comment (hdonnay) [reply httptransport/client/indexer.go:99]: hence ```case "", `application/json`:```

File: httptransport/notificationshandler.go (hunk 6 above)
Comment (hdonnay) [reply httptransport/notificationshandler.go:149]: It's called out in the headers.

I waffle on how useful reporting it via the trailer is, but this is at least centralizing handling the writer errors in one spot, one way.

File: httptransport/client/indexer.go (hunk 4 above)
Comment (ldelossa) [reply httptransport/client/indexer.go:99]: Why the switch case then? If you're catching the default case?

File: httptransport/updateoperationhandler.go (hunk 7 above)
Comment (hdonnay) [reply httptransport/updateoperationhandler.go:117]: I think so. Something flagged the useless returns for me, iirc

File: notifier/callback.go (hunk 9 above)
Comment (hdonnay) [reply notifier/callback.go]: Strictly necessary, no, but these methods are the only places where we don't get the piecewise encoding of the `codec` package "for free," so I thought it'd be prudent to minimize the number of intermediary allocations here.

Comment (ldelossa) [reply notifier/callback.go]: Do you think you loose the majority of this PRs effectiveness if we moved away from writing out the json directly?

File: httptransport/client/indexer.go (hunk 3 above)
Comment (ldelossa) [reply httptransport/client/indexer.go:41]: Misread that

File: httptransport/client/indexer.go (hunk 4 above)
Comment (hdonnay) [reply httptransport/client/indexer.go:99]: To return an error if other content-types are specified

File: notifier/callback.go (hunk 9 above)
Comment (hdonnay) [reply notifier/callback.go]: The majority? No, it just seemed like low-hanging fruit.

**Comment by jan-zmeskal** [comment]
@hdonnay @ldelossa I obviously don't want to rush this PR as I support your drive to do it right. I was just thinking if we could get 5c1b27fba36973620b41de8bc1f3fec2d811f09e merged separately, as many people seem to run into this issue on the main branch.

File: notifier/callback.go (hunk 9 above)
Comment (ldelossa) [reply notifier/callback.go]: Id prefer to not do this, string building json json is pretty ugly, and not great to maintain. As long as we increase the majority of the efficiency, we can still keep maintainability in mind. Amdahl's law comes to mind.

File: httptransport/client/indexer.go (hunk 4 above)
Comment (ldelossa) [reply httptransport/client/indexer.go:99]: Right, so im saying i dont think we should restrict it at all, we have a JSON api, try to parse the JSON and return an error if invalid. Is there any harm in not being defensive here for convenience ?

**Comment by hdonnay** [comment]
regarding the pipes, the JSON encoder is doing some buffering, so it's not like every object is incurring channel overhead.

Jan reported memory consumption dropping >10x in a bad instance; I'm okay with paying some cycles to get that sort of memory pressure reduction.

**Review by ldelossa — APPROVED** [review]

**Review by ldelossa — APPROVED** [review]
